File: c++/Heap&HeapSort/Heap.cpp
#include <iostream>
#include "Heap.h"

using namespace std;

void swap(int *x, int *y)
{
    int temp = *x;
    *x = *y;
    *y = temp;
}

/* CONSTRUCTOR */
Heap::Heap(int capac)
{
    this->capacity = capac;
    this->size = 0;
    this->arr = new int[capac];
}

/**
 * Function to heapify from an index (sift a value down)
 * 
 * @param index			The index to start from
 */
void Heap::maxHeapify(int index)
{
    // Get the right and left children
    int left = index * 2;
    int right = index * 2 + 1;

    // Set the initial largest to the index
    int largest = index;

    // Get the largest value of the two children and the index
    if(left <= this->size && this->arr[left] > this->arr[largest])
        largest = left;
    
    if(right <= this->size && this->arr[right] > this->arr[largest])
        largest = right;

    // If the largest is not the index, swap the index and the largest and recursively call to heapify
    if(largest != index)
    {
        swap(this->arr[index], this->arr[largest]);
        maxHeapify(largest);
    }
}

/**
 * Heapify Function for the Function "heapSort". Uses root at index 0 instead of 1 like usual
 * 
 * @param arr				The array to heapify
 * @param index				The index to heapify from
 * @param size				The size of the heap
 */
void Heap::maxHeapify(int *arr, int index, int size)
{
    int left = 2 * index + 1;
    int right = 2 * index + 2;
    int largest;

    // If the left child is larger than the index, set the largest to the left child index
	if(left < size && arr[left] > arr[index])
		largest = left;
	// Else, the index is the largest
	else
		largest = index;
				
	// Check if the right child is larger than the largest 
	if(right < size && arr[right] > arr[largest])
		largest = right;

    if(largest != index)
    {
        swap(arr[index], arr[largest]);
        maxHeapify(arr, largest, size);
    }
}

/**
 * Function to insert a value into the heap
 * 
 * @param value				The value to insert
 */
void Heap::insert(int value)
{
    if(this->size == this->capacity - 1)
    {
        cout << "Could not insert, heap full" << endl;
        return;
    }

    this->size++;
    int index = this->size;

    this->arr[index] = value;
    siftUp(index);

    cout << "Inserted value " << value << " into the heap!" << endl;
}

/**
 * Function to sift a value up from an index
 * 
 * @param index				The index to sift up from
 */
void Heap::siftUp(int index)
{
    while(index != 1 && this->arr[index / 2] < this->arr[index])
    {
        swap(this->arr[index], this->arr[index / 2]);
        index = index / 2;
    }
}

/**
 * Function to remove and return the max value in the heap
 * 
 * @return			The removed max number 
 */
int Heap::removeMax()
{
    if(isEmpty())
    {
        cout << "Heap empty" << endl;
        return -1;
    }

    if(this->size == 1)
    {
        this->size--;
        return this->arr[1];
    }

    int num = this->arr[1];

    this->arr[1] = this->arr[this->size];
    this->size--;
    maxHeapify(1);

    return num;
}

/**
 * Method that takes in an unsorted array and sorts it using in-place heapsort
 * 
 * @param array				Unsorted aray to be sorted
 * @param n                 The size of the array
 */
void Heap::heapSort(int *array, int n)
{
    for(int i = n / 2 - 1; i >=0; i--)
        maxHeapify(array, i, n);

    for(int i = n - 1; i >= 0; i--)
    {
        swap(array[0], array[i]);
        maxHeapify(array, 0, i);
    }
}
/**
 * Method to print the heap
 */
void Heap::printHeap()
{
    for(int i = 1; i <= this->size; i++)
    {
        cout << this->arr[i] << " ";
    }
    cout << endl;
}

int main()
{
    Heap *myHeap = new Heap(10);

    myHeap->insert(45);
	myHeap->insert(65);
	myHeap->insert(5);
	myHeap->insert(189);
	myHeap->printHeap();
	cout << "Removed the max from the heap: " << myHeap->removeMax() << endl;
	myHeap->printHeap();
	cout << "Removed the max from the heap: " << myHeap->removeMax() << endl;
	myHeap->printHeap();
		
	int myArray[] = {45, 12, 67, 123, 46};
		
	cout << "Unsorted array before heap-sort" << endl;
	for(int i = 0; i < 5; i++)
	{
		cout << myArray[i] << " ";
	}
	cout << endl;
		
	myHeap->heapSort(myArray, 5);
		
	cout << "Sorted array after heap-sort" << endl;
	for(int i = 0; i < 5; i++)
	{
		cout << myArray[i] << " ";
	}
	cout << endl;

    return 0;
}

File: java/Heap&Heapsort/Heap.java
/**
 * Class for a heap. Methods implemented are using a max-heap
 * 
 * @author MPinheiro
 *
 */
public class Heap {
	
	private int[] arr;
	private int capacity;
	private int size;
	
	/* CONSTRUCTOR */
	public Heap(int capac)
	{
		this.capacity = capac;
		this.size = 0;
		this.arr = new int[capac];
	}
	
	/**
	 * Method to heapify from an index (sift a value down)
	 * 
	 * @param index			The index to start from
	 */
	public void maxHeapify(int index)
	{
		// Get the right and left children of the node
		int left = 2 * index;
		int right = 2 * index + 1;
		int largest;
		
		// If the left child is larger than the index, set the largest to the left child index
		if(left <= this.size && this.arr[left] > this.arr[index])
			largest = left;
		// Else, the index is the largest
		else
			largest = index;
		
		// Check if the right child is larger than the largest 
		if(right <= this.size && this.arr[right] > this.arr[largest])
			largest = right;
		
		// If the largest value is NOT the index, swap the two values and recursively call the method to keep sifting down
		if(largest != index)
		{
			int temp = this.arr[index];
			this.arr[index] = this.arr[largest];
			this.arr[largest] = temp;
			maxHeapify(largest);
		}
	}
	
	/**
	 * Heapify method for the method "heapSort". Uses root at index 0 instead of 1 like usual
	 * 
	 * @param arr				The array to heapify
	 * @param index				The index to heapify from
	 * @param size				The size of the heap
	 */
	private void maxHeapify(int[] arr, int index, int size)
	{
		// Get the right and left children of the node
		int left = 2 * index + 1;
		int right = 2 * index + 2;
		int largest;
				
		// If the left child is larger than the index, set the largest to the left child index
		if(left < size && arr[left] > arr[index])
			largest = left;
		// Else, the index is the largest
		else
			largest = index;
				
		// Check if the right child is larger than the largest 
		if(right < size && arr[right] > arr[largest])
			largest = right;
				
		// If the largest value is NOT the index, swap the two values and recursively call the method to keep sifting down
		if(largest != index)
		{
			int temp = arr[index];
			arr[index] = arr[largest];
			arr[largest] = temp;
			maxHeapify(arr, largest, size);
		}
	}
	
	/**
	 * Method to insert a value into the heap
	 * 
	 * @param value				The value to insert
	 */
	public void insert(int value)
	{
		// If the heap is full
		if(this.size == capacity - 1)
		{
			System.out.println("Could not insert, heap full");
			return;
		}
		
		// Update the size, insert the value at the end and sift up
		this.size++;
		int index = this.size;
		this.arr[index] = value;		
		siftUp(index);		
		
		System.out.println("Inserted value " + value + " into the heap!");
	}
	
	/**
	 * Method to sift a value up from an index
	 * 
	 * @param index				The index to sift up from
	 */
	public void siftUp(int index)
	{
		// Keep sifting up until no longer needed
		while(index != 1 && this.arr[index / 2] < this.arr[index])
		{
			// Swap the elements and update the index so it points to the parent
			int temp = this.arr[index];
			this.arr[index] = this.arr[index / 2];
			this.arr[index / 2] = temp;
			index = index / 2;
		}
	}
	
	/**
	 * Method to remove and return the max value in the heap
	 * 
	 * @return			The removed max number 
	 */
	public int removeMax()
	{
		// If the heap is empty
		if(isEmpty())
		{
			System.out.println("Heap Empty");
			return -1;
		}
		
		// If the heap is of size 1
		if(this.size == 1)
		{
			this.size--;
			return this.arr[1];
		}
		
		
		// Save the number to return
		int num = arr[1];
		// Set the root equal to the last leaf node
		arr[1] = arr[this.size];
		// Decrease size
		this.size--;
		// Heapify's the heap. Sifts the value down
		maxHeapify(1);
		
		return num;
	}
	
	/**
	 * Method that takes in an unsorted array and sorts it using in-place heapsort
	 * 
	 * @param array				Unsorted aray to be sorted
	 */
	public void heapSort(int[] array)
	{
		// Build heap
		for(int i = array.length / 2 - 1; i >= 0; i--)
			maxHeapify(array, i, array.length);
		
		// Go through the heap, removing max and heapifying it again
		for(int i = array.length - 1; i >= 0; i--)
		{
			int temp = array[0];
			array[0] = array[i];
			array[i] = temp;
			
			maxHeapify(array, 0, i);
		}
			
	}
	
	/**
	 * Method to return but not remove the max value
	 * 
	 * @return
	 */
	public int getMax()
	{
		return this.arr[1];
	}
	
	/**
	 * Method to print the heap
	 */
	public void printHeap()
	{
		for(int i = 1; i <= this.size; i++)
		{
			System.out.print(this.arr[i] + " ");
		}
		System.out.println();
	}
	
	/**
	 * Method to get the size of the heap
	 * 
	 * @return				The size
	 */
	public int getSize()
	{
		return this.size;
	}
	
	/**
	 * Method to see if the heap is empty
	 * 
	 * @return				True if the heap is empty
	 */
	public boolean isEmpty()
	{
		return (this.size == 0);
	}

}
